spec/requests/halls_spec.rb

```diff
@@ -0,0 +1,345 @@
+      end
+    end
+
+    context "when user without permission" do
+      before { sign_in user }
+      it "redirects to root" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response).to redirect_to("/")
+      end
+
+      it "returns redirect status" do
+        patch("/halls/#{hall.id}")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user with permission" do
+      before { sign_in manager }
+      it "returns redirect status" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response.status).to eq(302)
+      end
+
+      it "redirects to /halls" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response).to redirect_to("/halls")
+      end
+
+      it "updates hall record" do
+        expect { patch("/halls/#{hall.id}", params:) }.to change { hall.reload.name }.from(hall.name).to(name.to_s)
+      end
+    end
+
+    context "when params invalid" do
+      before { sign_in manager }
+      let(:duration) { "s" }
+
+      it "doesn't create hall record" do
+        expect { patch("/halls/#{hall.id}", params:) }.not_to(change { hall.reload.name })
+      end
+
+      it "returns unsuccessful response" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response.status).to eq(422)
+      end
+
+      let(:duration) { Faker::Number.number(digits: 2) }
+      let(:name) { nil }
+
+      it "doesn't create hall record" do
+        expect { patch("/halls/#{hall.id}", params:) }.not_to(change { hall.reload.name })
+      end
+
+      it "returns unsuccessful response" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response.status).to eq(422)
+      end
+    end
+  end
+
+  describe "DELETE /halls" do
+    let(:hall) { create :hall }
+    context "when no user" do
+      it "redirects to sign_in" do
+        delete("/halls/#{hall.id}")
+        expect(response).to redirect_to("/users/sign_in")
+      end
+
+      it "returns redirect status" do
+        delete("/halls/#{hall.id}")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user without permission" do
+      before { sign_in user }
+      it "redirects to root" do
+        delete("/halls/#{hall.id}")
+        expect(response).to redirect_to("/")
+      end
+
+      it "returns redirect status" do
+        delete("/halls/#{hall.id}")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user with permission" do
+      let(:admin) { create :user, email: "testadmin@test.com", role: 2 }
+      before { sign_in admin }
+      it "returns redirect status" do
+        delete("/halls/#{hall.id}")
+        expect(response.status).to eq(302)
+      end
+
+      it "redirects to /halls" do
+        delete("/halls/#{hall.id}")
+        expect(response).to redirect_to("/halls")
+      end
+
+      it "destroy hall record" do
+        expect { delete("/halls/#{hall.id}") }.not_to change(Hall, :count)
```

Comment:
Should be:
```suggestion
        expect { delete("/halls/#{hall.id}") }.to change(Hall, :count)
```
as it is now it check if the hall was NOT deleted

I would personally do it like this:
```suggestion
        delete("/halls/#{hall.id}"
        expect (Hall.where(id: hall.id).count).to eq(0)
```